private/tools/java/com/github/bazelbuild/rules_jvm_external/maven/MavenPublisher.java

```diff
@@ -65,21 +65,35 @@ public class MavenPublisher {
     "file:/", "https://", "gs://", "s3://", "artifactregistry://"
   };
 
+  private static String getArg(String[] args, String envName, int argPos) {
+    String envVal = System.getenv(envName);
+    if (envVal != null) {
+      return envVal;
+    }
+    if (argPos < args.length) {
+      return args[argPos];
+    }
+    return null;
+  }
+
   public static void main(String[] args)
       throws IOException, InterruptedException, ExecutionException, TimeoutException {
-    String repo = args[0];
+    String repo = getArg(args, "MAVEN_REPO", 0);
     if (!isSchemeSupported(repo)) {
       throw new IllegalArgumentException(
           "Repository must be accessed via the supported schemes: "
               + Arrays.toString(SUPPORTED_SCHEMES));
     }
 
-    boolean gpgSign = Boolean.parseBoolean(args[1]);
-    Credentials credentials = new BasicAuthCredentials(args[2], args[3]);
+    boolean gpgSign = Boolean.parseBoolean(getArg(args, "GPG_SIGN", 1));
+    Credentials credentials = new BasicAuthCredentials(
+            getArg(args, "MAVEN_USER", 2),
+            getArg(args, "MAVEN_PASSWORD", 3)
+    );
 
     List<String> parts = Arrays.asList(args[4].split(":"));
```

Comment:
If the user isn't passing in the user or password on the command line, then the coordinates are likely to be at args position 2, right? Or is it okay because the script it quoting them, so we end up with empty values?

Reply:
Yes, the quoted empty string in `_TEMPLATE` in `maven_publish. bzl` will ensure the arg positions are consistent even if empty.